Engine/main/game_run.cpp

```diff
@@ -910,30 +920,72 @@ void SetupLoopParameters(int untilwhat,long udata,int mousestuff) {
     restrict_until=untilwhat;
     user_disabled_data=udata;
     user_disabled_for=FOR_EXITLOOP;
-    return;
 }
 
 // This function is called from lot of various functions
 // in the game core, character, room object etc
-void GameLoopUntilEvent(int untilwhat,long daaa) {
+static void GameLoopUntilEvent(int untilwhat,const void* daaa) {
   // blocking cutscene - end skipping
   EndSkippingUntilCharStops();
 
   // this function can get called in a nested context, so
   // remember the state of these vars in case a higher level
   // call needs them
-  int cached_restrict_until = restrict_until;
-  int cached_user_disabled_data = user_disabled_data;
-  int cached_user_disabled_for = user_disabled_for;
+  auto cached_restrict_until = restrict_until;
+  auto cached_user_disabled_data = user_disabled_data;
+  auto cached_user_disabled_for = user_disabled_for;
 
-  SetupLoopParameters(untilwhat,daaa,0);
+  SetupLoopParameters(untilwhat,daaa);
   while (GameTick()==0);
 
+  our_eip = 78;
+
   restrict_until = cached_restrict_until;
   user_disabled_data = cached_user_disabled_data;
   user_disabled_for = cached_user_disabled_for;
 }
 
+void GameLoopUntilValueIsZero(const char *value) 
+{
+    GameLoopUntilEvent(UNTIL_CHARIS0, value);
```

Comment:
how does going through one more level of indirection increase type safety ?

Reply:
The first function demands particular type, and nested one supplies correct enum constant for that type.